Client/core/CDiscordManager.cpp

```diff
@@ -0,0 +1,345 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto
+ *  LICENSE:     See LICENSE in the top level directory
+ *  FILE:        Client/core/CDiscordManager.cpp
+ *
+ *  Multi Theft Auto is available from http://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+
+#include "StdInc.h"
+#include "SharedUtil.Thread.h"
+#include <chrono>
+
+CDiscordManager::CDiscordManager() : m_discordCore(nullptr), m_bSuicide(false), m_waitingForServerName(false), m_storedActivity{}
+{
+    Reconnect(true); // Try to interact with discord on construction
+    m_pThread = new CThreadHandle(CDiscordManager::DiscordThread, this);
+
+    m_storedActivity.GetAssets().SetLargeImage("multi-theft-auto"); // Always thing
+    m_storedActivity.GetAssets().SetLargeText("Playing MTA:SA");
+}
+
+CDiscordManager::~CDiscordManager()
+{
+    m_bSuicide = true;
+    int iTries = 0;
+    while (iTries++ < 400 && m_bSuicide) // Wait maximum of 2 sec on this
+        Sleep(5);
+    if (m_bSuicide) m_pThread->Cancel(); // Kill it anyway
+    delete m_pThread;
+    SAFE_DELETE(m_discordCore);
+}
+
+// Establish connection with discord
+void CDiscordManager::Reconnect(bool bOnInitialization)
+{
+    bool discordRichPresence;
+    CVARS_GET("discord_rich_presence", discordRichPresence);
+    if (!discordRichPresence) return; // Disabled
+
+    m_threadSafety.lock();
+    discord::Result res = discord::Core::Create(DISCORD_CLIENT_ID, DiscordCreateFlags_NoRequireDiscord, &m_discordCore);
+    m_threadSafety.unlock();
+
+    if (!m_discordCore && bOnInitialization) // Output error only when trying to connect on initialization
+        WriteErrorEvent(SString("[DISCORD]: Failed to instantiate core, error code: %i", static_cast<int>(res)));
+    else
+    {
+        m_discordCore->SetLogHook(discord::LogLevel::Info, DiscordLogCallback);
+        WriteDebugEvent(SString("[DISCORD]: Instantiated at %08X", m_discordCore));
+
+        m_discordCore->ActivityManager().RegisterCommand(SString("%s /from_discord", *GetParentProcessPathFilename(GetCurrentProcessId())));
+        m_discordCore->ActivityManager().OnActivityJoin.Connect(OnActivityJoin);
+
+        if (!bOnInitialization) // Player could be in a server or in menu by now
+        {
+            if (g_pCore->IsConnected()) Restore();
+            else g_pCore->ResetDiscordRichPresence();
+        }
+    }
+}
+
+void CDiscordManager::DiscordLogCallback(discord::LogLevel level, const char* message)
+{
+    SString strMessage("[DISCORD]: %s", message);
+    if (level >= discord::LogLevel::Warn) WriteDebugEvent(strMessage);
+    else WriteErrorEvent(strMessage);
+}
+
+void CDiscordManager::OnActivityJoin(const char* joinSecret)
+{
+    if (!joinSecret)
+        return;
+
+    // Parse it
+    SString ipport, secret;
+    ushort  port;
```

Comment:
Better declare and define these variables in a single statement (to reduce the scope).